### ISSUE TITLE
Inconsistencies in the Husformer Paper and Code

### ISSUE THREAD

**Issue by Bluemax666** [SMARTlab-Purdue/Husformer#10]
Hello,

I have studied the Husformer paper and the associated code, and I've noticed some inconsistencies. If you're reading this, you might be attempting to reproduce the paper's results.

I also tried, and my conclusion is that I wasn't able to observe any significant score differences between Husformer and simpler neural network architectures like MLP or LSTM for the datasets CogLoad, MOCAS and WESAD. I cannot post the specific code I used for my reproduction attempt because the original Husformer code does not appear to have an open-source license (e.g., MIT) that would permit modification and redistribution of derivative works.

However, here are the points I'd like to highlight regarding the paper and the available code:

In the paper (https://arxiv.org/pdf/2209.15182):

- The formula for accuracy in multi-class classification appears incorrect. In Section IV.D, it states that the accuracy for the *i*-th class is: accuracy<sub>i</sub> = (TP<sub>i</sub> + TN<sub>i</sub>) / length<sub>i</sub>,  where length<sub>i</sub> is "the number of total samples in the i-th class." A simple example illustrates the issue: Imagine a classifier for 3 classes (A, B, C), with 5 samples for each class. If the classifier predicts everything correctly, using this formula, we would have: accuracy<sub>A</sub> = (TP<sub>A</sub> + TN<sub>A</sub>) / length<sub>A</sub>  = (5<sub>TP<sub>A</sub></sub> + 10<sub>TN<sub>A</sub></sub>) / 5<sub>length<sub>A</sub></sub> = 3. However, an accuracy value should typically be between 0 and 1.

- The paper presents the framework as "end-to-end," but in Section III.F, Algorithm 1, the computations for the initial layers (temporal convolutions and positional encoding, steps 6-7) appear to be outside the main training loop (steps 8-18). Why is this the case if it's truly end-to-end, where all parameters would typically be learned within the iterative training process?

- At the end of Section IV.A, it's mentioned that the input data used in the experiments are 1-second segments. This raises several questions:
-- The paper states multiple times that transformers are used to model "long-term interactions" between modalities. How can a 1-second window be considered "long-term" in this context?
-- For the CogLoad dataset (Table IV), given the sampling rates (mostly 1Hz), a 1-second segment might correspond to an input vector with very few values (e.g., potentially 6 values if using HR, IBI, GSR, SKT, and 2 ACC channels, all at 1Hz). How can there be such significant differences in accuracies (as shown in Table VI, Section IV) between the different methods tested on the CogLoad dataset with such potentially sparse input per segment?

In the code :

- Section III.F of the paper (Equation 9) mentions that the network's output goes through a softmax function, and the paper clearly frames the tasks as classification problems. Therefore, the final output of the model should be a probability distribution over the classes.
In the provided code (e.g., main-5.py), why does the neural network is set to have an output dimension of 1? Furthermore, in eval_metrics.py, the metrics appear to be designed to receive scalar values, and no softmax operation seems to be applied before computing these metrics for classification evaluation. This contradicts the paper's description.

- In data/cogload.pkl the modality names are EEG, GSR, BVP and POW when you load the data but if you look at the values they seem to correspond to the modalities IBI, GSR, HR and ACC.

- There are other problems with the code but they mostly come from fact that they took the code from https://github.com/yaohungt/Multimodal-Transformer and didn't change it too much

**Comment by shui27345-jpg** [comment]
> Hello,
> 
> I have studied the Husformer paper and the associated code, and I've noticed some inconsistencies. If you're reading this, you might be attempting to reproduce the paper's results.
> 
> I also tried, and my conclusion is that I wasn't able to observe any significant score differences between Husformer and simpler neural network architectures like MLP or LSTM for the datasets CogLoad, MOCAS and WESAD. I cannot post the specific code I used for my reproduction attempt because the original Husformer code does not appear to have an open-source license (e.g., MIT) that would permit modification and redistribution of derivative works.
> 
> However, here are the points I'd like to highlight regarding the paper and the available code:
> 
> In the paper (https://arxiv.org/pdf/2209.15182):
> 
> * The formula for accuracy in multi-class classification appears incorrect. In Section IV.D, it states that the accuracy for the _i_-th class is: accuracyi = (TPi + TNi) / lengthi,  where lengthi is "the number of total samples in the i-th class." A simple example illustrates the issue: Imagine a classifier for 3 classes (A, B, C), with 5 samples for each class. If the classifier predicts everything correctly, using this formula, we would have: accuracyA = (TPA + TNA) / lengthA  = (5TPA + 10TNA) / 5lengthA = 3. However, an accuracy value should typically be between 0 and 1.
> * The paper presents the framework as "end-to-end," but in Section III.F, Algorithm 1, the computations for the initial layers (temporal convolutions and positional encoding, steps 6-7) appear to be outside the main training loop (steps 8-18). Why is this the case if it's truly end-to-end, where all parameters would typically be learned within the iterative training process?
> * At the end of Section IV.A, it's mentioned that the input data used in the experiments are 1-second segments. This raises several questions:
>   -- The paper states multiple times that transformers are used to model "long-term interactions" between modalities. How can a 1-second window be considered "long-term" in this context?
>   -- For the CogLoad dataset (Table IV), given the sampling rates (mostly 1Hz), a 1-second segment might correspond to an input vector with very few values (e.g., potentially 6 values if using HR, IBI, GSR, SKT, and 2 ACC channels, all at 1Hz). How can there be such significant differences in accuracies (as shown in Table VI, Section IV) between the different methods tested on the CogLoad dataset with such potentially sparse input per segment?
> 
> In the code :
> 
> * Section III.F of the paper (Equation 9) mentions that the network's output goes through a softmax function, and the paper clearly frames the tasks as classification problems. Therefore, the final output of the model should be a probability distribution over the classes.
>   In the provided code (e.g., main-5.py), why does the neural network is set to have an output dimension of 1? Furthermore, in eval_metrics.py, the metrics appear to be designed to receive scalar values, and no softmax operation seems to be applied before computing these metrics for classification evaluation. This contradicts the paper's description.
> * In data/cogload.pkl the modality names are EEG, GSR, BVP and POW when you load the data but if you look at the values they seem to correspond to the modalities IBI, GSR, HR and ACC.
> * There are other problems with the code but they mostly come from fact that they took the code from https://github.com/yaohungt/Multimodal-Transformer and didn't change it too much

Hello, based on your response to another question, I modified the Husformer code to use multi-class average accuracy and multi-class average F1-score (using the MOCAS dataset). However, my results are significantly better than those reported in the original paper. Could your reproduced code accurately replicate the paper's accuracy metrics?

**Comment by Bluemax666** [comment]
Yes i get mostly higher scores than in the paper, here are my results, (I used a downsampled version of MOCAS dataset)

<img width="601" height="228" alt="Image" src="https://github.com/user-attachments/assets/f52b3268-01f9-46c2-b8b0-377fcc5e1950" />

**Comment by shui27345-jpg** [comment]
> Yes i get mostly higher scores than in the paper, here are my results, (I used a downsampled version of MOCAS dataset)
> 
> <img alt="Image" width="601" height="228" src="https://private-user-images.githubusercontent.com/44173031/479256360-f52b3268-01f9-46c2-b8b0-377fcc5e1950.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU2MDQ4MjUsIm5iZiI6MTc1NTYwNDUyNSwicGF0aCI6Ii80NDE3MzAzMS80NzkyNTYzNjAtZjUyYjMyNjgtMDFmOS00NmMyLWI4YjAtMzc3ZmNjNWUxOTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODE5VDExNTUyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjYzUxZGY0MTY0Yjc3MjgzNDZmMDI2NmMyMzdhNjI4NjY0YjE0YWNiMGU4NmE3NjVhYTYwZDRkNjU1OWJjYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z01zfr4yD1ja9qo08u4F1F87GQF3NB0ajS3Cbz8PxW8">

I also used this dataset, but the accuracy rate after five rounds of training reached 99%. Did you make major changes to the original code structure? Is there a significant difference between the original code structure and the algorithm structure in the paper? I modified the output to a three-dimensional vector, did not use the softmax layer, and before calculating the accuracy rate, I converted the results to the position of the maximum probability, changing 0 to -1, and then calculated the accuracy rate. The accuracy rate is surprisingly high. I basically didn't make other changes.

**Comment by shui27345-jpg** [comment]
> Yes i get mostly higher scores than in the paper, here are my results, (I used a downsampled version of MOCAS dataset)
> 
> <img alt="Image" width="601" height="228" src="https://private-user-images.githubusercontent.com/44173031/479256360-f52b3268-01f9-46c2-b8b0-377fcc5e1950.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU2NzQ4OTksIm5iZiI6MTc1NTY3NDU5OSwicGF0aCI6Ii80NDE3MzAzMS80NzkyNTYzNjAtZjUyYjMyNjgtMDFmOS00NmMyLWI4YjAtMzc3ZmNjNWUxOTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODIwVDA3MjMxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4ZTg2ODM3YmQ3MGIxMDljMDkzNTAwNWE2OTMwYTI4MDI2YmYyNjU5ZWM5YTU3ZGNjZjdjODNkZDk4YzFjYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mTMHgibtDHLUcvPBg5kY572606jK6TFb_-cJE6TJLro">

Hello, may I ask if your MOCAS dataset also contains a lot of duplicate data? Could you please share how you handled it? Can we connect on WeChat to discuss further？

**Comment by wiaobang** [comment]
> Yes i get mostly higher scores than in the paper, here are my results, (I used a downsampled version of MOCAS dataset)
> <img alt="Image" width="601" height="228" src="https://private-user-images.githubusercontent.com/44173031/479256360-f52b3268-01f9-46c2-b8b0-377fcc5e1950.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY5MjcyMTUsIm5iZiI6MTc2NjkyNjkxNSwicGF0aCI6Ii80NDE3MzAzMS80NzkyNTYzNjAtZjUyYjMyNjgtMDFmOS00NmMyLWI4YjAtMzc3ZmNjNWUxOTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjI4VDEzMDE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQwY2FlNDZkODkxZWIzYzI5MTVhY2JhNTJiNDAyMTdiNTM4MmFjMTZjYzk3MDU1OWViODA3NGI2NjY2MjBjMDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lqsU_ypmTAhPkSBX5Syjpwju0c54CPgTHWpl1nR11EM">

Hello, have you tested this Husformer algorithm on the DEAP dataset? The results I got were far worse than the accuracy mentioned in the paper. Thank you for your reply.

**Comment by git-boring** [comment]
Hello, I'd like to ask how you modified and got the code running. I want to run it on the DEAP dataset.

**Comment by Bluemax666** [comment]
> Hello, may I ask if your MOCAS dataset also contains a lot of duplicate data? Could you please share how you handled it? Can we connect on WeChat to discuss further？

To do the dataset i took only 1 line every 100 lines in the MOCAS datasheet

**Comment by Bluemax666** [comment]
> Hello, I'd like to ask how you modified and got the code running. I want to run it on the DEAP dataset.

I made the ouput dimension 3 for cogload, mocas and wesad (its 1 by default)

i used 
def remake_label(target):
    target = target.squeeze(-1)
    return target

i remade the f1 score and accuracies function

i also factorised the code so it works with any number of modalities

also alpha in focalloss has to be the same dimension as the output dimension
